### PR TITLE
#162949090 Article Bookmark Feature 

### PR DESCRIPTION
**Pull request by Teatoller**
### What does this PR do?
Allows a user to rate an article they have not authored.

### Description of Task to be completed?
- [x] Add tests
- [x] Create a feature to bookmark an article
- [x] Have the following endpoints working:
- [x] POST ```/api/v1/article/<slug>/bookmark/```
- [x] GET ```/api/v1/article/<slug>/bookmark/all/```
### How should this be manually tested?
1. Clone this repository on your local machine using 

    ```$```  ```git clone https://github.com/andela/ah-robotics.git```
2. Change your working directory to the project's directory using 
     
     ```$``` ```cd ah-robotics```
3. Change to the specific branch `ft-user-add-bookmark-162949090` using 

     ```$``` ```git checkout ft-user-add-bookmark-162949090```
4. Create a virtual environment using the following command: 
     ```$```  ```python3 -m virtualenv env```
5. Activate the virtual environment using 
```$``` ```source env/bin/activate```
6. Install all the requirements using 
     ```$``` ```pip install -r requirements.txt```

    #### Database setup

    Ensure [PostgreSQL](https://www.postgresql.org/) is installed and is running, follow these steps to create a new user and a database which will be used in the application.

    1. Start the postgres database server using the command:
      
       ```$``` ```pg_ctl -D /usr/local/var/postgres start```
    
    2. Run psql interactive terminal
    
        ```$``` ```psql postgres```
    
    3. Create a new user:
    
        ```$``` ```CREATE USER sample_username WITH PASSWORD 'sample_password';```
    
    4. Grant privileges to the user:
    
        ```$``` ```ALTER USER sample_username CREATEDB;```
    
    5. Create a database:
    
       ```$``` ```CREATE DATABASE sample_database_name WITH OWNER sample_username;```
    
    6. Exit psql interactive with the following command: ```$``` ```\q```
    
7. Create a **.env** file and add the following configuration variables

    ```source-json
          source env/bin/activate
          export DB_NAME="database_name"
          export DB_USER="sample_username"
          export DB_PASS="sample_password"
          export DB_HOST="host_name_eg_localhost"
          export DB_PORT="port_number_eg_5432"
          export TIME_DELTA="30"
          export EMAIL_HOST="smtp.gmail.com"
          export EMAIL_PORT=587
          export EMAIL_HOST_USER="user@example.com"
          export EMAIL_HOST_PASSWORD="email_password"
          export PASSWORD_RESET_URL="api/v1/reset_password"
          export CLIENT_DOMAIN="127.0.0.1:8000"
          export CLOUDINARY_NAME="your_cloudinary_name"
          export CLOUDINARY_KEY="your_cloudinary_key"
          export CLOUDINARY_SECRET="your_secret_key"
    ```
    
    > **Note**: The values of some of these variables are private namely the EMAIL_HOST_USER, EMAIL_HOST_PASSWORD and the CLOUDINARY VALUES.  Use your GMAIL email address and password for the  EMAIL_HOST_USER and EMAIL_HOST_PASSWORD respectively. To get cloudinary values, create an account with cloudinary and obtain the values.

8. Source .env to set the configuration variables in the terminal:

      ```$``` ```source .env```
9. Run migrations through the terminal to create the tables in the database:
      
      (i) Make the migrations using: 
      
      ```$``` ```python manage.py makemigrations```
      
      (ii) Apply the migrations using: 
      
      ```$``` ```python manage.py migrate```

10. Run the following command to collect static files.

   ```$``` ```python manage.py collectstatic```
      
#### Testing the features

1. Run the server from the terminal using: 
    
  ```$```  ```python manage.py runserver```
2. Download [Postman](https://www.getpostman.com/) and after it is installed, open it on your local machine and Paste the URL `http://127.0.0.1:8000/api/v1/users/` to first sign up a user.
3. On the **Headers** section add **Content-type** as the KEY and **_application/json_** as the VALUE.
4. Head over to the **Body** section and select the **raw** option and then add a payload to the body using the following format:

    ```
        {
            "user": {
                "username": "sample_username",
                "email": "sample_email",
                "password": "sample_password"
                }
        }
    ```

**Example** 
sample_username---> honululu, 
sample_email-->hanku@example.com, 
password-->test1234

5. Once all the fields are inserted, ensure that it is a **POST** request and click the **Send** button.
6. Log in the user using the following endpoint `http://127.0.0.1:8000/api/v1/users/login/` and insert the following payload to the body section.
    
    ``` {
            "user": {
                "email": "sample_email",
                "password": "sample_password"
                }
        }
    ```

**NB** Use details you created in preceding step ```4```
7. Once all the fields are inserted, ensure that it is a **POST** request and click the **Send** button.
8. After logging in the screen should look as follows:
    
    
screenshot 2019-01-29 at 15 59 01

9. Copy the token and click on the **Headers** section. In the **Key** column, type in the word **Authorization**. And in its corresponding value in the **Value** column, write the word __token__, add a space and paste in the token that you copied from the login response. The **Headers** section should look as follows after doing so:
    
    image

10. To Create an article, insert the following URL: http://127.0.0.1:8000/api/v1/articles/. In the body section, ensure the type is _raw_ and the format is _JSON(application/json)_. Finally, post an article by using the following payload:
    
    ```
    {
    "article": {
        "title": "Sample Article",
        "description": "Sample Article Description",
        "body": "The sample article body",
        "tagList": ["samples", "articles"]
    }
}
    ```
    Once all the fields are inserted, ensure that it is a **POST** request and click the **Send** button.

11. To get all the articles, use the URL http://127.0.0.1:8000/api/v1/articles/. Change the method to a **GET** and hit **Send**.

12. To get a specific article, Copy the slug and insert it at the end of the articles URL as follows:
 http://127.0.0.1:8000/api/v1/articles/slug/.  Change the method to a **GET** and hit **Send**. 
For a example, to get a specific article with **my-first-article** as the slug name, the URL would be as follows: http://127.0.0.1:8000/api/v1/articles/my-first-article/. 
    > **Note**: The slug is a field in the article response body

13. On the Postman change the PARAM url to ```http://127.0.0.1:8000/api/v1/articles/sample-article/bookmark/``` and make sure method is **POST** then click ```Send```

**Screenshot showing article "Bookmarked"**

screenshot 2019-01-31 at 14 27 13

**NB** Try to create and bookmark several items.

14. When you click on **Send** a second time it _removes the bookmark_ by a delete.

screenshot 2019-01-31 at 14 27 24

 
15. To see all **Bookmarked** article, paste in Postman PARAM the following URL   ```http://127.0.0.1:8000/api/v1/articles/bookmark/all/``` 

screenshot 2019-01-31 at 14 27 59

#### Running tests
1. Head over to the terminal and ensure that the server is not running and you are in the project root folder.
2. Run the tests by typing in the following command: 
    
    `python manage.py test`

### Any background context you want to provide?
None

### What are the relevant pivotal tracker stories?
[#162949090](https://www.pivotaltracker.com/n/projects/2235171)

## Questions:
Questions and comments are welcome.

### CODE REVIEW

File: authors/apps/bookmark/models.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
+from django.db import models
+from ..authentication.models import User
+from authors.apps.articles.models import Article
+from authors.apps.profiles.models import UserProfile
+from django.utils.text import slugify
```
Comment (johnwayodi) [inline authors/apps/bookmark/models.py:5]: This import seems like it is not being used in the file, consider removing it.

**Review by Paulvitalis200 — COMMENTED** [review]
Hey @Teatoller ,
This is good work. I have managed to test the feature locally on my machine and everything looks good. Good job.

File: authors/apps/bookmark/views.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
+from django.shortcuts import render
+from rest_framework.generics import CreateAPIView, ListAPIView
+from .serializers import BookmarkSerializer
+from authors.apps.articles.models import Article
+from authors.apps.articles.views import article_not_found
+from .models import BookmarkArticle
+from rest_framework.response import Response
+from rest_framework import status
```
Comment (johnwayodi) [inline authors/apps/bookmark/views.py:7]: This import doesn't seem to be used in the file, consider removing it.
> from django.shortcuts import render

Also consider reorganizing the imports and separate local from library imports.

**Review by johnwayodi — COMMENTED** [review]
Hello @Teatoller, I've however spotted some minimal issues and left some comments, kindly look into them.

**Comment by Teatoller** [comment]
> Hey @Teatoller ,
> This is good work. I have managed to test the feature locally on my machine and everything looks good. Good job.

Thank you @Paulstar200

File: authors/apps/bookmark/views.py (hunk 2 above)
Comment (Teatoller) [reply authors/apps/bookmark/views.py:7]: @johnwayodi ,
Thank you for picking this up.
I am actioning it right away.

Regards
Steven Ennis

**Pull request merged**